helix-core/src/main/java/org/apache/helix/participant/DistClusterControllerStateModel.java

```diff
@@ -85,7 +85,7 @@ public void onBecomeStandbyFromLeader(Message message, NotificationContext conte
 
     logger.info(controllerName + " becoming standby from leader for " + clusterName);
 
-    if (_controller != null) {
+    if (_controllerOpt.isPresent()) {
```

Comment:
If this is not in the scope of lock, could we leave leakage there when a race condition happen?

Reply:
1. State transitions are all happening in one thread. So there is no race condition between the reset here and the creation logic. This PR is to fix the NPE when the reset is triggered by shutdown. So this line won't cause the issue even without the fix.
2. This fix is only to fix NPE, it won't fix all potential leakage. Note that if the partition is in ERROR state, then when we drop the partition, there is no clean up. This is the most serious leakage, actually. If the test is still not stable, I plan to fix this problem in another PR.

Reply:
I remember Xiaoyuan's change will make the state model can call reset and ST at same time, right?

Reply:
That would be fine. Your concern is basically "onBecomeLeaderFromStandby" and "onBecomeStandbyFromLeader" are called at the same time. But even with Xiaoyuan's change, this is not possible, right?
If the reset is triggered by the other logic, like shutdown, then we have synchronized (_controllerOpt) in reset() to ensure the calls are mutually exclusive.